### ISSUE TITLE
Move type information from condition to actions

### ISSUE THREAD

**Issue by whiskeysierra** [zalando/riptide#62]
The current syntax looks like this

```java
Success success = rest.execute(GET, url)
        .dispatch(series(),
                on(SUCCESSFUL, Success.class).capture(),
                on(ACCEPTED, Success.class).map(this::handle).capture(),
                on(CLIENT_ERROR)
                    .dispatch(status(),
                            on(UNAUTHORIZED).call(this::login),
                            on(UNPROCESSABLE_ENTITY)
                                    .dispatch(contentType(),
                                            on(PROBLEM, ThrowableProblem.class).call(propagate()),
                                            on(ERROR, Exception.class).call(propagate()))),
                on(SERVER_ERROR)
                    .dispatch(statusCode(),
                            on(503).call(this::retryLater)),
                anySeries().call(this::fail))
        .as(Success.class).orElse(null);
```

I'm thinking whether it makes sense to move the type information from conditions to action level:

```java
Success success = rest.execute(GET, url)
        .dispatch(series(),
                on(SUCCESSFUL).capture(Success.class),
                on(ACCEPTED).map(Success.class, this::handle).capture(),
                on(CLIENT_ERROR)
                    .dispatch(status(),
                            on(UNAUTHORIZED).call(this::login),
                            on(UNPROCESSABLE_ENTITY)
                                    .dispatch(contentType(),
                                            on(PROBLEM).call(ThrowableProblem.class, propagate()),
                                            on(ERROR).call(Exception.class, propagate()))),
                on(SERVER_ERROR)
                    .dispatch(statusCode(),
                            on(503).call(this::retryLater)),
                anySeries().call(this::fail))
        .as(Success.class).orElse(null);
```

@lukasniemeier-zalando What do you think?

**Comment by tkrop** [comment]
I like this idea. Moves the type information where it is needed making the distinction between typed and untyped conditions unnecessary.

**Comment by ePaul** [comment]
Looks nice.

If all capturing branches have captures with same types (or types with some common supertype), the mapping in the end might also be easier (i.e. you don't need the `as` anymore, as the type is already there). (For this to work completely, we would need a null type (i.e. a type which is a subtype of everything) for the throwing branches, though).

**Comment by whiskeysierra** [comment]
> For this to work completely, we would need a null type (i.e. a type which is a subtype of everything) for the throwing branches, though

I get the idea, but I don't think it's possible with the type system in Java. The only thing that is a valid instance for every type is `null`, but there is no type we can use for those branches.

**Comment by whiskeysierra** [comment]
Fixed with #70
